akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -475,34 +516,52 @@ private[stream] final case class LimitWeighted[T](n: Long, costFn: T ⇒ Long) e
 /**
  * INTERNAL API
  */
-private[akka] final case class Sliding[T](n: Int, step: Int) extends PushPullStage[T, immutable.Seq[T]] {
-  private var buf = Vector.empty[T]
-
-  override def onPush(elem: T, ctx: Context[immutable.Seq[T]]): SyncDirective = {
-    buf :+= elem
-    if (buf.size < n) {
-      ctx.pull()
-    } else if (buf.size == n) {
-      ctx.push(buf)
-    } else if (step > n) {
-      if (buf.size == step)
-        buf = Vector.empty
-      ctx.pull()
-    } else {
-      buf = buf.drop(step)
-      if (buf.size == n) ctx.push(buf)
-      else ctx.pull()
+private[akka] final case class Sliding[T](n: Int, step: Int) extends GraphStage[FlowShape[T, immutable.Seq[T]]] {
+  require(n > 0, "n must be greater than 0")
+  require(step > 0, "step must be greater than 0")
+
+  val in = Inlet[T]("Sliding.in")
+  val out = Outlet[immutable.Seq[T]]("Sliding.out")
+  override val shape: FlowShape[T, immutable.Seq[T]] = FlowShape(in, out)
+
+  override protected val initialAttributes: Attributes = DefaultAttributes.sliding
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler with OutHandler {
+    private var buf = Vector.empty[T]
+
+    override def onPush(): Unit = {
+      buf :+= grab(in)
+      if (buf.size < n) {
+        pull(in)
+      } else if (buf.size == n) {
+        push(out, buf)
+      } else if (step <= n) {
+        buf = buf.drop(step)
+        if (buf.size == n) {
+          push(out, buf)
+        } else pull(in)
+      } else if (step > n) {
+        if (buf.size == step) {
+          buf = buf.drop(step)
+        }
+        pull(in)
+      }
     }
-  }
 
-  override def onPull(ctx: Context[immutable.Seq[T]]): SyncDirective =
-    if (!ctx.isFinishing) ctx.pull()
-    else if (buf.size >= n) ctx.finish()
-    else ctx.pushAndFinish(buf)
+    override def onPull(): Unit = {
+      pull(in)
+    }
```

Comment:
the original `onPull` and `onUpstreamFinish ` says:
- always `pull` upstream when context is not finishing
- push and finish if there is remaining in current window (ie. `buf.size < n && buf.size > 0`)
- finish if current window is empty (ie. `buf.size >= n || buf.size ==0`)

There is no necessary to introduce deferred completion again, and I think the code is ok, am I right?